src/contribute/testing.md

```diff
@@ -371,3 +371,146 @@ RunnableOnService tests are tests built to use the Beam TestPipeline class,
 which enables test authors to write simple functionality verification. They are
 meant to use some of the built-in utilities of the SDK, namely PAssert, to
 verify that the simple pipelines they run end in the correct state.
+
+
+### Effective use of the TestPipeline JUnit rule
+
+`TestPipeline` is JUnit rule designed to facilitate testing pipelines. 
+In combination with `PAssert`, the two can be used for testing and 
+writing assertions over pipelines. However, in order for these assertions 
+to be effective, the constructed pipeline **must** be run by a pipeline 
+runner. If the pipeline is not run (i.e., executed) then the 
+constructed `PAssert` statements will not be triggered, and will thus 
+be ineffective. 
+
+To prevent such cases, `TestPipeline` has some protection mechanisms in place.
+
+__Abandoned node detection (performed automatically)__
+
+Abandoned nodes are `PTransforms`, `PAsserts` included, that were not 
+executed by the pipeline runner. Abandoned nodes are most likely to occur 
+due to the one of the following scenarios:
+ 1. Lack of a `pipeline.run()` statement at the end of a test. 
+ 2. Addition of `PTransform`s  after the pipeline has already run.
+
+Abandoned node detection is *automatically enabled* when a real pipeline 
+runner (i.e. not a `CrashingRunner`) and/or a 
+`@NeedsRunner` / `@RunnableOnService` annotation are detected. 
+
+Consider the following test:
+  
+```java
+// Note the @Rule annotation here
+@Rule
+public final transient TestPipeline pipeline = TestPipeline.create();
+
+@Test
+@Category(NeedsRunner.class)
+public void myPipelineTest() throws Exception {
+
+final PCollection<String> pCollection = 
+  pipeline
+    .apply("Create", Create.of(WORDS).withCoder(StringUtf8Coder.of()))
+    .apply(
+        "Map1",
+        MapElements.via(
+            new SimpleFunction<String, String>() {
+
+              @Override
+              public String apply(final String input) {
+                return WHATEVER;
+              }
+            }));
+            
+PAssert.that(pCollection).containsInAnyOrder(WHATEVER);       
+
+/* ERROR: pipeline.run() is missing, PAsserts are ineffective */
+}
+```
+ 
+The `PAssert` at the end of this test method will not be executed, since 
+`pipeline` is never run, making this test ineffective. If this test method 
+is run using an actual pipeline runner, an exception will be thrown 
+indicating that there was no `run()` invocation in the test.
+
+Exceptions that are thrown prior to executing a pipeline, will fail 
+the test unless handled by an `ExpectedException` rule.
+
+Consider the following test:  
+  
+```java
```

Comment:
Due to Javascript issues (unrelated to this PR), it is possible that the browser doesn't render the code snippet if Python snippet was last selected on another page.

To work around the problem, I suggest to add an empty Python snippet below (see example in Kenn's state blog post).